### PR TITLE
Create a new KafkaClientSupplier on each REST request

### PR DESCRIPTION
**Pull request by spena**
### Description 
This PR allows an external dependency to pass specific configurations to each REST request in order to create a custom `KafkaClientSupplier`. 

This new `ConfiguredKafkaClientSupplier` is used to create the `ServiceContext` that is injected on each REST request.

A couple of new classes are created where these REST configurations can be set: `KsqlRestContextProvider` and `KsqlRestContext`. 
- The `KsqlRestContextProvider` initializes a default context in case there is no external dependency that initializes it. 
- The `KsqlRestContextProvider` can keep a REST configuration on the REST thread local scope. This allows the application to have different REST configurations per user.

The current `KsqlRestServiceContextBinder`, `KsqlRestServiceContextFactory`, and `DefaultServiceContext` classes are modified to create and add a `ConfiguredKafkaClientSupplier` to the `ServiceContext`.

### Testing done 
_Describe the testing strategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestContextProvider.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.context;
+
+import java.util.Collections;
+
+public final class KsqlRestContextProvider {
+  private static final ThreadLocal<KsqlRestContext> restContextThreadLocal = new ThreadLocal<>();
```
Comment (agavra) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestContextProvider.java]: This seems like a hacky... we shouldn't be using ThreadLocal as an API to communicate information (aside from the fact that ThreadLocal can be a dangerous abstraction). This also makes the APIs very confusing... I have no idea what's going on.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/ConfiguredKafkaClientSupplier.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2018 Confluent Inc.
```
Comment (rodesai) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/ConfiguredKafkaClientSupplier.java]: nit: year

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/ConfiguredKafkaClientSupplier.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+ * {@link KafkaClientSupplier} implementation that injects extra configurations to the kafka
+ * admin and consumer/producer clients.
+ */
+public class ConfiguredKafkaClientSupplier implements KafkaClientSupplier {
+  private final KafkaClientSupplier defaultSupplier;
+  private final ImmutableMap<String, Object> supplierProperties;
+
+  public ConfiguredKafkaClientSupplier(
+      final KafkaClientSupplier defaultSupplier,
+      final Map<String, Object> config
```
Comment (rodesai) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/ConfiguredKafkaClientSupplier.java]: nit: use the same name as the member

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestContextProvider.java (hunk 1 above)
Comment (rodesai) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestContextProvider.java]: I agree. Is there any guarantee filters and resources are called on the same thread?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestServiceContextFactory.java [hunk 4]
```diff
@@ -18,22 +18,48 @@
...
 
 /**
  * This class implements {@link Factory}, which allows a REST application to create
  * a new {@link ServiceContext} during REST requests.
  */
 public class KsqlRestServiceContextFactory implements Factory<ServiceContext> {
-  private final KsqlConfig ksqlConfig;
+  private static KsqlConfig ksqlConfig;
+
+  public static void configure(final KsqlConfig ksqlConfig) {
```
Comment (agavra) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestServiceContextFactory.java]: I think this should be syncrhonized

Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestServiceContextFactory.java]: even if it is a single variable assignment inside?

Comment (agavra) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestServiceContextFactory.java]: maybe not necessary... but it's basically free :)

Comment (rodesai) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestServiceContextFactory.java]: there's only one thread that uses it right?

Comment (spena) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestServiceContextFactory.java]: Yes, one single thread. It is initialized from the `KsqlRestApplication` only once.

**Pull request merged**